src/core/mesh/qgsmesheditor.h

```diff
@@ -0,0 +1,393 @@
+ email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSMESHEDITOR_H
+#define QGSMESHEDITOR_H
+
+#include <QObject>
+#include <QUndoCommand>
+#include <QPointer>
+
+#include "qgsmeshdataset.h"
+#include "qgsmeshdataprovider.h"
+#include "qgstriangularmesh.h"
+
+#ifndef SIP_RUN
+
+class QgsMeshVertexCirculator;
+
+#endif //SIP_RUN
+
+/**
+ *  \brief Class that wraps a QgsMesh to ensure the consistency of the mesh during edition and help to access to elements from other elements
+ *
+ *  A topologic face need to:
+ *  - be convex
+ *  - counter clock wise
+ *  - not share an unique vertex with another face
+ */
+class CORE_EXPORT QgsTopologicMesh
+{
+  public:
+
+    enum TopologicErrorType
+    {
+      NoError,
+      InvalidFace,
+      FlatFace,
+      UniqueSharedVertex,
+      InvalidVertex
+    };
+
+    struct TopologicError
+    {
+      TopologicError(): errorType( NoError ), elementIndex( -1 ) {}
+      TopologicError( TopologicErrorType type, int elementIndex ): errorType( type ), elementIndex( elementIndex ) {}
+      TopologicErrorType errorType;
+      int elementIndex;
+
+      bool operator==( const TopologicError &other ) const {return ( other.errorType == errorType && other.elementIndex == elementIndex );}
+      bool operator!=( const TopologicError &other ) const {return !operator==( other );}
+    };
+
+#ifndef SIP_RUN
+
+    using FaceNeighbors = QVector<int>;
+
+    /**
+     *  Class that contains independant faces an topological information about this faces
+     */
+    class TopologicFaces
+    {
+      public:
+        QVector<QgsMeshFace>  meshFaces() const {return mFaces;}
+
+        void clear();
+
+      private:
+        QVector<QgsMeshFace> mFaces; // the faces containing the vertices indexes in the mesh
+        QVector<FaceNeighbors> mFacesNeighborhood; // neighborhood of the faces, face indexes are local
+        QHash<int, int> mVerticesToFace; // map of vertices to incident face, face indexes are local
+        QList<int> mBoundaries; // list of boundary vertices indexes in the mesh
+
+        friend class QgsTopologicMesh;
+        friend class QgsMeshVertexCirculator;
+    };
+
+    //! Class that contains topologic differences between two states of a topologic mesh, only accessible from the QgsTopologicMesh class
+    class TopologicChanges
```

Comment:
this is class with all member private and just friend class? maybe better to convert to struct and move to "private" section?

Reply:
also seeing this, I suggest to create a common patter like public class QgsTopologicMesh in the QGIS_CORE api and then private class qgstopologicalmesh_p.[h/cpp] with the actual private implementation where you can have such helper constructs.